---
title: "native_late_adipoDE"
author: "Sarah Hazell Pickering"
date: "2/9/2022"
output: html_document
---

## Summary

The D7 floating adipocytes are quite different between replicates, but still very distinct from bulk adipocyte samples (see PCA plot).  I was a bit concerned about the normalisation of some of the donor7 replicates (samples 418 and 419 which were both P8 I think).  But in the end I settled for a stringent pvalue threshold of 0.01 and the GO terms make alot of sense.  Floating adipocytes are enriched for mitochondrial metabolism genes or adipogenesis hallmarks, whilst bulk samples are enriched for genes related to the extracellular matrix and/or blood vessel morphogenesis. 
  
outputs are rpkm tables (as always) and native_bulk.v.float.tab, which shows pvalues and logFCs for DE between bulk and floating adipocytes (D7 only and D13 only stats are given in the right most columns as well).


```{r}
library(limma)
library(biomaRt)
library(org.Hs.eg.db)
library(ggplot2)
library(RUVSeq)
library(clusterProfiler)
limma_dir = "/projects/imb-pkbphil/sp/rnaseq/late_adipo_feb22/output/03limma"
```


### Load read counts
```{r}
fc = read.delim(file.path(limma_dir, "../02featureCounts/late_adipo_feb22.counts"),
                header=T, stringsAsFactors = F, skip = 1)
colnames(fc) = gsub("output.01hisat.","", gsub(".sorted.bam","",colnames(fc)))
head(fc)
```
### Load sample information
```{r}
sample_info = read.delim(file.path(limma_dir, "late_adipo_sample_info.txt"),
                header=T)
sample_info$sample_id = gsub("-",".", sample_info$sample_id)
sample_info$bio.condition = gsub(" ",".", sample_info$bio.condition)
sample_info
```

###Combine into DGE object
```{r}
ob = DGEList(counts = data.matrix(fc[7:ncol(fc)]),
             samples = sample_info,
             group = sample_info$bio.condition,
             genes = fc[c(1,6)])
rownames(ob$counts) = ob$genes$Geneid
```

### Filter out apex samples
```{r}
ob$counts = ob$counts[,ob$samples$cell_type =="native"]
ob$samples = ob$samples[ob$samples$cell_type =="native",]
head(ob)
```

## PCA

Bulk and floating transcriptomes are more different for donor7 than donor13.  This first dimension does separate bulk and floating (though better for donor7) and the second dimension seems to separate donors.  Bulk adipocytes have similarities between donors, but the floating adipocytes seem quite different from one another, even with donor7 there's a lot of variability.  

```{r}
mds = plotMDS(ob, labels = ob$samples$group, top=5000)
```

## Filtering

```{r}
plotRLE(ob$counts, outline=FALSE, col=ob$samples$group, main="Before Filtering")

plotSmear(ob, main = "Before Filtering")

nrow(ob) #58 735 genes before filtering
```


```{r}
filt = ob[filterByExpr(ob),, keep.lib.sizes=FALSE]
nrow(filt) #17 882 genes kept after filtering
```

The exact filter is approxiamtely 10/ libsize in millions (now about 0.2 CPM).  This is a lower filter than previous, I guess because our library sizes are bigger.  
```{r}
10/(median(ob$samples$lib.size)/1000000) #0.190 minimum CPM 
```

Hmm interesting there's higher expression in the floating samples, especially for lower CPM, higher sensitivity perhaps?  
```{r}
plotSmear(filt, main = "After Filtering, D07G") 
plotSmear(filt, main = "After Filtering, D13A", pair=c("D13A","D13A.floating")) 
plotRLE(filt$counts, outline=FALSE, 
        col=filt$samples$treat, main="After filtering")
```

Sanity check that the geneids haven't been mixed up.  
```{r}
filt$genes[filt$genes$Geneid == "ENSG00000228630",] #Hotair is still included

filt$genes[filt$genes$Geneid == "ENSG00000132170",] #PPARG is included
```
## Normalise

Hmm these libraries have quite a bit of substructure, the means are not around zero, and the variances have quite large differences, particularly for two of those donor7 floating samples.  You can see they sit differently in the PCA.  

```{r}
filt = calcNormFactors(filt,method = "TMM")
filt$samples[1:5]
plotRLE(cpm(filt$counts, normalized.lib.sizes =FALSE), outline=FALSE, 
             main="After TMM normalisation (no libsize)")

plotRLE(cpm(filt$counts, normalized.lib.sizes = TRUE), outline=FALSE, 
        col=filt$samples$group, main="After TMM normalisation (with libsize)")
```

## Harsher filter
mincount = 20 keeps 16009 genes
15104 kept at min count 30 
14003 kept at min count 50

if we set the group to separation method (so donor7 and 13 are pooled); and filt at least 50 counts we get 13254 kept genes

Doesn't change the RLE plot and you can still see bias in the smears, though the extremely low expression genes are removed.  Have to try something a bit more drastic to normalise these properly (RUV for example), but we can get an estimate of the DE.  

```{r}
ob$samples$group = ob$samples$separation
harsh = ob[filterByExpr(ob, min.count=50),, keep.lib.sizes=FALSE]
nrow(harsh) #14003 kept at min count 50 

plotSmear(harsh, main = "After Filtering, D07G") 
#plotSmear(harsh, main = "After Filtering, D13A", pair=c("D13A","D13A floating")) 
#plotSmear(harsh, main = "After Filtering, D13A", pair=c("D07G floating","D13A floating")) 

plotRLE(harsh$counts, outline=FALSE, 
        col=harsh$samples$group, main="After filtering")
```

```{r}
harsh = calcNormFactors(harsh,method = "TMM", refColumn = 12)
harsh$samples[1:5]
plotRLE(cpm(harsh$counts, normalized.lib.sizes =FALSE), outline=FALSE, 
             main="After TMM normalisation (no libsize)", col=filt$samples$group)

plotRLE(cpm(filt$counts, normalized.lib.sizes = TRUE), outline=FALSE, 
        col=filt$samples$group, main="After TMM normalisation (with libsize)")
```

## Annnotate with gene names

```{r}
mart <- biomaRt::useMart(biomart = "ensembl",
  dataset = "hsapiens_gene_ensembl",
  host = "jan2019.archive.ensembl.org")
annot = getBM(c("external_gene_name", "description", "ensembl_gene_id", "gene_biotype"), 
              filters = "ensembl_gene_id",
              values = as.character(filt$genes$Geneid),
              mart = mart,
              useCache = F)
head(annot)

#Tidying up the annot table
annot$description = gsub("\\[Source:.+\\]", "", annot$description)
colnames(annot)[1] = "gene_name"
colnames(annot)[3] = "gene"

#Add gene names to data object
filt$genes = merge(filt$genes, annot, by.x= "Geneid", 
                        by.y = "gene", sort=FALSE)
colnames(filt$genes) = c("gene","length","gene_name","description","gene_biotype")
head(filt$genes)
```

### usually make FPKM tables here
but I'm going to hold off until I can check out the normalisation

## Limma DE

Well the limma mean variance trend looks fine :)

```{r}
filt$samples$group = factor(filt$samples$group)
norm_ob = voom(filt, plot=T)
```

### Simple separate design

Hmm that's a lot of different genes, quite the lenient test.  I don't trust this with the normalisation differences.  Oh and when we set the threshold lower donor13 effect is gone, and the total effect is driven by the donor7.  

```{r}
design = model.matrix(~0+norm_ob$targets$group)
colnames(design) = levels(norm_ob$targets$group)
design

tests = makeContrasts(D7.bulk.v.float = D07G.floating - D07G,
                       D13.bulk.v.float = D13A.floating - D13A,
                      bulk.v.float = (D07G.floating - D07G) + (D13A.floating - D13A),
                      D7.v.D13 = D07G - D13A,
                      floating.D7.v.D13 = D07G.floating - D13A.floating,
                       levels = design) #test for which genes change the same 

simple_fit = lmFit(norm_ob, design=design)
simple_fit = contrasts.fit(simple_fit, tests)
simple_fit = eBayes(simple_fit, robust=T)

results= decideTests(simple_fit, method="separate", adjust.method = "BH")
summary(results)

summary(decideTests(simple_fit, method="separate", adjust.method = "BH", p.value = 0.01))
vennDiagram(results)
vennDiagram(results[,3:5])
```

If we adjust the pvalues as if this was one test, we spread the significant genes more evenly between the conditions.  It's worth being a little more stringent on the pvalue threshold as well, considering how many D7 genes are showing up.  

This is not a bad test, we have 213 genes that are found in all three tests and so super robust.  Only two genes seem to be significant in opposite directions which is a good sign.  Interestingly the combined test picks up 563 genes that were ignored in the individual tests.  There are less genes found uniquely in each individual test than in the combined, which I also take as a good sign.  

```{r}
stringent = decideTests(simple_fit, method="global", adjust.method = "BH", p.value = 0.01)
summary(stringent)

vennDiagram(stringent[,3:5])
vennDiagram(stringent[,1:3])
```

## Make a combined table

In an ANOVA-style combined test 4.5k genes are different between bulk and floating (not shown). Instead I opted for the combined test, which presents 4.1k DE genes at a 0.01 pvalue threshold.  

Overwhelmingly significant genes are downregulated in floating cells, which makes sense with the loss of population diversity.  

Top gene preferentially expressed in bulk samples is MEG3, a potential lncRNA tumor supressor, via proliferation control.   Another top gene is WEE1, a negative regulator of mitosis entry.  There is also CDH11.  These genes are also pretty highly expressed.  

Genes preferentially expressed in floating adipocytes include a lot of TCA cycle and metabolism genes.  Though not in the top 10 galectin 12 is expressed by floating adipocytes, as well as some other galectins.   

```{r}
common = topTable(simple_fit, coef = "bulk.v.float", number=nrow(simple_fit$genes))
head(common[3:ncol(common)], n=10) #bulk genes

head(common[common$logFC > 0,3:ncol(common)], n=10)

common[grep("galectin", common$description),]

summary(common$adj.P.Val < 0.01)
```

Add some information about the individual tests as well?

```{r}
d7 = topTable(simple_fit, coef = "D7.bulk.v.float", number=nrow(simple_fit$genes))

d13 = topTable(simple_fit, coef = "D13.bulk.v.float", number=nrow(simple_fit$genes))

both = merge(d7, d13, suffixes=c("_d7", "_d13"), by=c("gene","length","gene_name","description", "gene_biotype"))
head(both)


all = merge(common, both, by=c("gene","length","gene_name","description", "gene_biotype"))
head(all)

all = all[grep("^(t.*|B_d.*|P.Value_d.*)", colnames(all), invert=T)]
all = all[order(all$adj.P.Val, -(all$B)),]
head(all)
```

```{r eval=F}
write.table(all, file.path(limma_dir, "native_bulk.v.float.tab"), 
            sep="\t", row.names=F, quote=F)
```

## Quick GO

```{r}
molsig <- clusterProfiler::read.gmt("/projects/imb-pkbphil/sp/annotations/molsigdb/msigdb.v7.4.symbols.gmt")
head(molsig); nrow(molsig)
##Pick gene sets to look in bcos all in a lot
prefixes = c("HALLMARK", "KEGG","REACTOME", "WP", "GOBP", "GOCC","GOMF")
colnames(molsig) = c("term","gene")
some.molsig = molsig[gsub("_.*","", molsig$term) %in% prefixes,]
some.molsig$term = factor(some.molsig$term)
table(gsub("_.*","", some.molsig$term))
```

Floating adipocytes are largely enriched for mitochondrial metabolism genes, but also 89 adipogenesis genes.  

```{r}
go_up = enricher(gene=common$gene_name[common$logFC > 0 &
                                      common$adj.P.Val < 0.01],
              TERM2GENE = some.molsig,
              minGSSize = 10, 
              universe = common$gene_name) #universe all expressed genes
print(head(go_up, n=10)[c(3:7)])

go_up[grep("ADIPO", go_up$ID), 3:9]

dotplot(go_up, showCategory=20)+ scale_y_discrete(label=function(x) tolower(stringr::str_trunc(x,45))) + 
theme(axis.text.y = element_text(size=10)) 
```

Bulk adipocytes are enriched for a variety of terms, with extracellular matrix very strong (likely driving the epithelial to mesenchymal gene set) as well as blood vessel and connective tissue GO terms.  In fact ~8% of significant protein coding genes are related to blood vessel morphogenesis (GeneRatio = 0.08).  

```{r}
go_down = enricher(gene=common$gene_name[common$logFC < 0 &
                                      common$adj.P.Val < 0.01],
              TERM2GENE = some.molsig,
              minGSSize = 10, 
              universe = common$gene_name) #universe all expressed genes
print(head(go_down, n=10)[c(3:7)])
barplot(go_down, showCategory=20)+ scale_x_discrete(label=function(x) tolower(stringr::str_trunc(x,45))) + 
theme(axis.text.y = element_text(size=10)) 
dotplot(go_down, showCategory=20)+ scale_y_discrete(label=function(x) tolower(stringr::str_trunc(x,45))) + 
theme(axis.text.y = element_text(size=10)) 

emapplot(go_down, showCategory = 10)

go_down[grep("MESENCHYMAL", go_down$ID), 3:9]
go_down[grep("EXTRACELLULAR", go_down$ID), 3:9]
go_down[grep("BLOOD_VESSEL", go_down$ID), 3:9]
```

```{r}
common[grep("SNAI", common$gene_name),]
```


## Volcanoplots

(-logfc is more expression in bulk adipocytes)

```{r}
volcanoplot(simple_fit, coef = "bulk.v.float", highlight=8, names = simple_fit$genes$gene_name)
volcanoplot(simple_fit, coef = "D7.bulk.v.float", highlight=2, names = simple_fit$genes$gene_name)
volcanoplot(simple_fit, coef = "D13.bulk.v.float", highlight=8, names = simple_fit$genes$gene_name)

```


## Check p-value distributions

Very strong effect for d7 and combined.  d13 looks robust as well.  
```{r}
ggplot(common) + geom_histogram(aes(x=P.Value), binwidth=0.01)

ggplot(d7) + geom_histogram(aes(x=P.Value), binwidth=0.01)

ggplot(d13) + geom_histogram(aes(x=P.Value), binwidth=0.01)
```



## Make FPKM table

```{r}
lib_rpkm = data.frame(rpkm(filt, normalize.lib.sizes=TRUE))
colnames(lib_rpkm) = paste(filt$samples$bio.condition, filt$samples$rep, sep="_")
lib_rpkm = lib_rpkm[order(colnames(lib_rpkm))]

format_rpkm = merge(filt$genes, lib_rpkm,  
                    by.x="gene", by.y = 'row.names',sort=FALSE)
head(format_rpkm)
```

Oh cool the normalisation is much nicer for the rpkm values :) Still high vraince for samples 13 and 14.  
```{r}
plotRLE(data.matrix(format_rpkm[6:ncol(format_rpkm)]), outline=FALSE)  #check normalisation
```

```{r eval=F}
write.table(format_rpkm, sep='\t', row.names = FALSE,quote = F,
            file=file.path(limma_dir,"native_rpkm_tmm.tab"))
```

```{r}
filt$samples$group = factor(filt$samples$group)
group_rpkm = data.frame(rpkmByGroup(filt,normalize.lib.sizes=TRUE))
head(group_rpkm)

colSums(group_rpkm)

format_grpkm = merge(filt$genes, group_rpkm,  
                    by.x="gene", by.y = 'row.names',sort=FALSE)
head(format_grpkm)
plotRLE(data.matrix(format_grpkm[6:ncol(format_grpkm)]), outline=FALSE)
```

```{r}
write.table(format_grpkm, sep='\t', row.names = FALSE, quote=F,
            file=file.path(limma_dir,"native_rpkm_tmm_means.tab"))
```